lightningd/payalgo.c

```diff
@@ -122,14 +123,49 @@ static void json_pay_failure(struct pay *pay,
 	command_fail_detailed(pay->cmd, r->errorcode, data, "%s", msg);
 }
 
+/* Determine if we should delay before retrying. Return a reason
+ * string, or NULL if we will not retry */
+static const char *delay_before_retry(const tal_t *ctx,
+				      const struct sendpay_result *r)
+{
+	/* The routing failures WIRE_EXPIRY_TOO_FAR, WIRE_EXPIRY_TOO_SOON,
+	 * and WIRE_FINAL_EXPIRY_TOO_SOON may arise due to disagreement
+	 * between the peers about what the block heights are. So
+	 * delay for those before retrying. */
+	if (!r->succeeded && r->errorcode == PAY_TRY_OTHER_ROUTE) {
+		switch (r->routing_failure->failcode) {
+		case WIRE_EXPIRY_TOO_FAR:
+		case WIRE_EXPIRY_TOO_SOON:
+		case WIRE_FINAL_EXPIRY_TOO_SOON:
+			return tal_fmt(ctx,
+				       "Possible blockheight disagreement "
+				       "(%s from peer)",
+				       onion_type_name(r->routing_failure->failcode));
+
+		default:
+			/* Do nothing */ ;
+		}
+	}
+
+	return NULL;
+}
+
 /* Start a payment attempt. */
 static bool json_pay_try(struct pay *pay);
 
+/* Used when delaying. */
+static void v_pay_try(struct pay *pay)
```

Comment:
What's the `v_` prefix stand for?

Reply:
`void` return. Sorry, ran out of ideas...